lua31/doc/luac.man
.\" $Id: luac.man,v 1.1 2000/07/20 06:57:22 jeske Exp $
.TH LUAC 1 "23 December 1997"
.SH NAME
luac \- Lua compiler
.SH SYNOPSIS
.B luac
[
.B \-c | -u
]
[
.B \-d
]
[
.B \-l
]
[
.B \-p
]
[
.B \-q
]
[
.B \-v
]
[
.B \-o
.I outputfile
]
.I sourcefile
\&.\|.\|.
.SH DESCRIPTION
.B luac
is the Lua compiler.
It translates programs written in the Lua programming language
into binary files that can be loaded and executed with
.B lua_dofile
in C or
.B dofile
in Lua.

.BR luac
produces a single output file containing the bytecodes
for all named source files.
By default,
the output file is named
.BR luac.out ,
but you can change this with the
.B \-o
option.

You can use "-" to indicate stdin as a source file.

.BR luac
can also load and list binary files with the
.B \-u
option.

.SH OPTIONS
.LP
.TP 1i
.B \-c
compile (this is the default).
.TP
.B \-u
undump, i.e., load and list the given binary files.
If no files are given, then luac undumps
.BR luac.out .
.TP
.BI \-D " name"
predefine symbol
.IR name
for conditional compilation.
.TP
.B \-d
turns debugging on.
.TP
.B \-l
produces a listing of the compiled bytecode for Lua's virtual machine.
This is the default when undumping.
.TP
.B \-p
parses sources files but does not generate any output file.
.TP
.B \-q
quiet; procudes no listing. This is the default when compiling.
.TP
.B \-v
prints version information.
.TP
.B \-V
verbose.
.TP
.BI \-o " outputfile"
name the output file
.IR output ,
instead of the default
.BR luac.out .
The output file cannot be a source file.
.SH FILES
.PD 0
.TP 20
.B luac.out
default output file
.PD
.SH "SEE ALSO"
.BR lua (1)
.LP
.IR "Reference Manual of the Programming Language Lua"
.LP
http://www.tecgraf.puc-rio.br/lua/
.SH DIAGNOSTICS
Error messages should be self explanatory.
.SH BUGS
Inherits any bugs from Lua,
but Lua has no bugs...
.SH AUTHORS
W. Celes, R. Ierusalimschy & L. H. de Figueiredo
(lua@tecgraf.puc-rio.br)


lua31/history
This is version 3.1 (alpha).

* Changes from version 3.0 to 3.1 (alpha)
  + NEW FEATURE: anonymous functions with closures (via "upvalues").
  + new syntax:
    - local variables in chunks.
    - better scope control with DO block END.
    - constructors can now be also written: { record-part; list-part }.
    - more general syntax for function calls and lvalues, e.g.:
      f().x=1
      o:f(x,y):g(z)
  + major code re-organization and clean-up; reduced module interdependecies.
  + support for multiple contexts.
  + new traversal functions "foreach" and "foreachvar".
  + changing lua to use doubles instead of floats is now trivial.
  + complete debug information stored in pre-compiled chunks.

* Changes from version 2.5 to 3.0
  + NEW CONCEPT: "tag methods".
    Tag methods replace fallbacks as the meta-mechanism for extending the
    semantics of Lua. Whereas fallbacks had a global nature, tag methods
    work on objects having the same tag (e.g., groups of tables).
    Existing code that uses fallbacks should work without change.
  + new, general syntax for constructors {[exp] = exp, ... }.
  + support for handling variable number of arguments in functions (varargs).
  + support for conditional compilation ($if ... $else ... $end).
  + cleaner semantics in API simplifies host code.
  + better support for writing libraries (auxlib.h).
  + better type checking and error messages in the standard library.
  + luac can now also undump.

* Changes from version 2.4 to 2.5
  + io and string libraries are now based on pattern matching;
    the old libraries are still available for compatibility
  + dofile and dostring can now return values (via return statement)
  + better support for 16- and 64-bit machines
  + expanded documentation, with more examples

* Changes from version 2.2 to 2.4
  + external compiler creates portable binary files that can be loaded faster
  + interface for debugging and profiling
  + new "getglobal" fallback
  + new functions for handling references to Lua objects
  + new functions in standard lib
  + only one copy of each string is stored
  + expanded documentation, with more examples

* Changes from version 2.1 to 2.2
  + functions now may be declared with any "lvalue" as a name
  + garbage collection of functions
  + support for pipes

* Changes from version 1.1 to 2.1
  + object-oriented support
  + fallbacks
  + simplified syntax for tables
  + many internal improvements


lua31/src/luac/luac.c
/*
** $Id: luac.c,v 1.1 2000/07/20 06:57:22 jeske Exp $
** lua compiler (saves bytecodes to files; also list binary files)
** See Copyright Notice in lua.h
*/

#include <stdarg.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "luac.h"
#include "lparser.h"
#include "lzio.h"
#include "luadebug.h"

void PrintChunk(TProtoFunc* Main);
void DumpChunk(TProtoFunc* Main, FILE* D);

static FILE* efopen(char* name, char* mode);
static void doit(int undump, char* filename);

static int listing=0;			/* list bytecodes? */
static int dumping=1;			/* dump bytecodes? */
static int undumping=0;			/* undump bytecodes? */
static int parsing=0;			/* parse only? */
static int verbose=0;			/* tell user what is done */
static FILE* D;				/* output file */

static void usage(void)
{
 fprintf(stderr,"usage: "
 "luac [-c | -u] [-D name] [-d] [-l] [-p] [-q] [-v] [-o output] file ...\n"
 " -c\tcompile (default)\n"
 " -u\tundump\n"
 " -D\tpredefine symbol for conditional compilation\n"
 " -d\tgenerate debugging information\n"
 " -l\tlist (default for -u)\n"
 " -o\toutput file for -c (default is \"luac.out\")\n"
 " -p\tparse only\n"
 " -q\tquiet (default for -c)\n"
 " -v\tshow version information\n"
 " -V\tverbose\n"
 );
 exit(1);
}

#define	IS(s)	(strcmp(argv[i],s)==0)

int main(int argc, char* argv[])
{
 char* d="luac.out";			/* default output file */
 int i;
 lua_open();
 for (i=1; i<argc; i++)
 {
  if (argv[i][0]!='-')			/* end of options */
   break;
  else if (IS("-"))			/* use stdin */
   break;
  else if (IS("-c"))			/* compile (and dump) */
  {
   dumping=1;
   undumping=0;
  }
  else if (IS("-D"))			/* $define */
  {
   TaggedString* s=luaS_new(argv[++i]);
   s->u.globalval.ttype=LUA_T_NUMBER;
   s->u.globalval.value.n=1;
  }
  else if (IS("-d"))			/* debug */
   lua_debug=1;
  else if (IS("-l"))			/* list */
   listing=1;
  else if (IS("-o"))			/* output file */
   d=argv[++i];
  else if (IS("-p"))			/* parse only (for timing purposes) */
  {
   dumping=0;
   parsing=1;
  }
  else if (IS("-q"))			/* quiet */
   listing=0;
  else if (IS("-u"))			/* undump */
  {
   dumping=0;
   undumping=1;
   listing=1;
  }
  else if (IS("-v"))			/* show version */
   printf("%s  %s\n(written by %s)\n\n",LUA_VERSION,LUA_COPYRIGHT,LUA_AUTHORS);
  else if (IS("-V"))			/* verbose */
   verbose=1;
  else					/* unknown option */
   usage();
 }
 --i;					/* fake new argv[0] */
 argc-=i;
 argv+=i;
 if (dumping || parsing)
 {
  if (argc<2) usage();
  if (dumping)
  {
   for (i=1; i<argc; i++)		/* play safe with output file */
    if (IS(d)) luaL_verror("will not overwrite input file \"%s\"\n",d);
   D=efopen(d,"wb");			/* must open in binary mode */
  }
  for (i=1; i<argc; i++) doit(0,IS("-")? NULL : argv[i]);
  if (dumping) fclose(D);
 }
 if (undumping)
 {
  if (argc<2)
   doit(1,"luac.out");
  else
   for (i=1; i<argc; i++) doit(1,IS("-")? NULL : argv[i]);
 }
 return 0;
}

static void do_compile(ZIO* z)
{
 TProtoFunc* Main=luaY_parser(z);
 if (listing) PrintChunk(Main);
 if (dumping) DumpChunk(Main,D);
}

static void do_undump(ZIO* z)
{
 while (1)
 {
  TProtoFunc* Main=luaU_undump1(z);
  if (Main==NULL) break;
  if (listing) PrintChunk(Main);
 }
}

static void doit(int undump, char* filename)
{
 FILE* f= (filename==NULL) ? stdin : efopen(filename, undump ? "rb" : "r");
 ZIO z;
 if (filename==NULL) filename="(stdin)";
 zFopen(&z,f,filename);
 if (verbose) fprintf(stderr,"%s\n",filename);
 if (undump) do_undump(&z); else do_compile(&z);
 if (f!=stdin) fclose(f);
}

static FILE* efopen(char* name, char* mode)
{
 FILE* f=fopen(name,mode);
 if (f==NULL)
 {
  fprintf(stderr,"luac: cannot open ");
  perror(name);
  exit(1);
 }
 return f; 
}
